### PR TITLE
Allow user to define the default compiler optimsation level

### PR DESCRIPTION
**Pull request by ocaisa**

### CODE REVIEW

File: easybuild/tools/options.py [hunk 1]
```diff
@@ -213,6 +213,9 @@ def override_options(self):
             'cleanup-builddir': ("Cleanup build dir after successful installation.", None, 'store_true', True),
             'cleanup-tmpdir': ("Cleanup tmp dir after successful run.", None, 'store_true', True),
             'color': ("Allow color output", None, 'store_true', True),
+            "default-optimisation": ("Set default optimization level, overriding easybuild optimization default "
+                                     "(possibilities are: noopt, lowopt, defaultopt, opt)",
```
Comment (boegel) [inline easybuild/tools/options.py]: hmm, we shouldn't hardcode the options here, we should try and obtain them via `easybuild.tools.toolchain.compiler`?

also, the option parser supports a `choice` type of option, so do:

```python
'default-opt-level': ("Specify default optimisation level", 'choice', 'store', 'defaultopt', known_opt_levels),
```

I like how you use different spelling for `optimisation` vs `optimization` in the option name and the help, so let's use `opt` instead ;)

File: easybuild/tools/toolchain/compiler.py [hunk 2]
```diff
@@ -231,9 +231,19 @@ def _set_compiler_flags(self):
         fflags = [self.options.option(x) for x in self.COMPILER_F_FLAGS + self.COMPILER_F_UNIQUE_FLAGS \
                   if self.options.get(x, False)]
 
+        # Allow a user-defined default optimisation
+        if build_option('default_optimisation'):
+            if build_option('default_optimisation') in self.COMPILER_OPT_FLAGS:
+                default_opt = self.options.option(build_option('default_optimisation'))
```
Comment (boegel) [inline easybuild/tools/toolchain/compiler.py]: you can simplify this a bit by only calling out to `build_option` once:

```python
default_opt_level = build_option('default_opt_level')
if default_opt_level in self.COMPILER_OPT_FLAGS:
    default_opt = default_opt_level
else:
    raise ...
```

Note that the `default_opt` default setting has just moved into `options.py` (see my other remark), no need to hardcode it anymore here

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/tools/options.py (hunk 1 above)
Comment (boegel) [inline easybuild/tools/options.py]: better:

```
from easybuild.tools.toolchain.compiler import Compiler
...
    'default-opt-level': ("Specify default optimisation level", 'choice', 'store', 'defaultopt', Compiler.COMPILER_OPT_FLAGS),
```

Comment (ocaisa) [inline easybuild/tools/options.py]: `opt` it is, copy paste is the enemy

File: easybuild/tools/toolchain/compiler.py [hunk 3]
```diff
@@ -231,9 +231,17 @@ def _set_compiler_flags(self):
...
                   if self.options.get(x, False)]
 
+        # Allow a user-defined default optimisation
+        default_opt_level = build_option('default_opt_level')
+        if default_opt_level in self.COMPILER_OPT_FLAGS:
+            default_opt = default_opt_level
+        else:
+            raise EasyBuildError("Unknown value for default optimisation: %s (possibilities are %s)" %
+                                 (default_opt_level, self.COMPILER_OPT_FLAGS))
+
```
Comment (ocaisa) [inline easybuild/tools/toolchain/compiler.py]: Invalid option already picked up by the option parser...could remove the if/else?

Comment (boegel) [inline easybuild/tools/toolchain/compiler.py]: Hmm, leave it there, it's pretty cheap, and we may change things later such that `default_opt_level` comes from somewhere else

File: easybuild/tools/toolchain/compiler.py [hunk 4]
```diff
@@ -231,9 +231,17 @@ def _set_compiler_flags(self):
         fflags = [self.options.option(x) for x in self.COMPILER_F_FLAGS + self.COMPILER_F_UNIQUE_FLAGS \
                   if self.options.get(x, False)]
 
+        # Allow a user-defined default optimisation
+        default_opt_level = build_option('default_opt_level')
+        if default_opt_level in self.COMPILER_OPT_FLAGS:
+            default_opt = default_opt_level
+        else:
+            raise EasyBuildError("Unknown value for default optimisation: %s (possibilities are %s)" %
+                                 (default_opt_level, self.COMPILER_OPT_FLAGS))
```
Comment (boegel) [inline easybuild/tools/toolchain/compiler.py]: how about

```python
default_opt_level = build_option('default_opt_level')
if default_opt_level not in self.COMPILER_OPT_FLAGS:
    raise ...
```

and then also use `default_opt_level` below

**Comment by boegel** [comment]
@ocaisa: fix for the failing tests in https://github.com/ocaisa/easybuild-framework/pull/28

**Comment by boegel** [comment]
Going in, thanks a lot @ocaisa!

**Pull request merged**